feat(keybroker-server): Use Rego to appraise EARs

thomas-fossati opened the pull request:
Use Rego to implement the appraisal policy for attestation results.

A default policy is hardcoded (see keybroker-server/src/policy.rego).

The user may provide their own policy to override the default using the `--policy` command line switch.

The user must provide a JSON file containing one or more RIM reference values.

Fix #13

Arnaud-de-Grandmaison-ARM commented on rust-keybroker/keybroker-server/src/verifier.rs:
```diff
@@ -60,14 +63,13 @@ pub fn verify_with_veraison_instance(
 
-    Ok(verified)
+    // Appraise the received EAR using the embedded policy (see ./policy.rego)
+    // unless a custom one has been provided on the command line.  The deafault
```
Typo: deafault

thomas-fossati replied:
```suggestion
    // unless a custom one has been provided on the command line.  The default
```

Arnaud-de-Grandmaison-ARM reviewed (approved):
I'm not familiar with that part of the code base, but it overall makes perfect sense to me from what I could understand from it.

paulhowardarm commented on rust-keybroker/keybroker-server/src/main.rs:
```diff
@@ -167,6 +173,14 @@ struct Args {
+
+    /// File containing a JSON array with base64-encoded known-good RIM values
+    #[arg(long, default_value = "rims.json")]
+    rims: String,
```
I wonder if we should consider a different name for this argument rather than `rims`, and also document it as being "JSON input data for the policy evaluation" to generalise it?

There are currently a few CCA-specific shortcuts in the internals of the key broker, but I've tried to avoid making it explicitly CCA-specific in terms of its documented behaviour or its command-line interface. I'd like to retain that agnosticism if we can, so that community contributors could use it to stand up equivalent demos on other types of CC platform.

Also, if I replace the REGO file with a custom one (which is supported), the input data might need to be of different structure, not just a RIM array. So it feels to me like the Rust code and the CLI should be decoupled from the expectations of the policy file.

One final point about this: the server and client both have a "mock" mode (where the server uses a static challenge value, and the client replays a hard-coded demo CCA token, allowing them to be tested on off-the-shelf dev platforms without any sort of TEE hardware or access to evidence). I think it would be good to extract the RIMs from the mock token and give the server some special knowledge about them, so that the mock mode continues to work out-of-the-box without requiring additional admin steps to create the required JSON file.

paulhowardarm commented on rust-keybroker/keybroker-server/src/policy.rs:
```diff
@@ -0,0 +1,75 @@
+    // Set the EAR claims-set to be appraised
+    engine.set_input(Value::from_json_str(ear_claims)?);
+
+    let results = engine.eval_rule("data.arm_cca.allow".to_string())?;
```
Again, just curious as to how this line would work if I override the policy with a custom REGO file. Is it correct that the package still needs to be `arm_cca` and the rule still needs to be `allow`? (Speaking as a Regorus/rego newbie here).

thomas-fossati replied:
At the moment the calling convention is:

1. The policy package name is `"arm_cca"`
2. The policy output is conveyed through the `"allow"` variabile
3. The result of the policy evaluation is either `"true"` or `"false"`
4. The policy takes "some" reference values as input (*)

(*) When using the default policy, (4) requires a precise scheme:
```cddl
start = {
  "reference-values": [ + b64-rim ]
}

b64-rim = text .b64c rim

rim = bytes .size 32
```

I agree we could make the whole thing more flexible.

For example, we can allow customising the OPA rule, which is currently hardcoded to `"arm_cca.allow"` adding a new CLI switch (e.g., `--rule <...>`).
This would relax (1) and (2).

Regarding (4) we must:

* Document what is the expectation for the default policy, and (maybe)
* Provide a custom policy/ref-vals example.

I think (3) should stay as is, which is pretty reasonable.

paulhowardarm commented on rust-keybroker/keybroker-server/src/policy.rs:
```diff
@@ -0,0 +1,75 @@
+// Evaluate an EAR claims-set against the appraisal policy and known-good RIM values
+pub(crate) fn rego_eval(
+    custom_policy: Option<String>,
+    rims: &str,
```
As with my comment on the CLI parameter name, I think something like `policy_data` or `reference_values` or something like that might be better than `rims` here.

thomas-fossati replied:
I like `reference_values`.

paulhowardarm reviewed (commented):
Really love this - great to see the key broker used as a platform to showcase AR appraisal policies. I had a few inline comments about whether we can relax the tie to CCA-specific concepts in order to leave the door open for other evidence profiles, and to make sure that there is a low-friction path to keeping the "mock modes" working without additional admin steps.

thomas-fossati commented:
Summary:
* Changed rims into reference-values
* Removed the ability to override the default policy
* Associate policy to evidence dynamically (only Arm CCA is supported for now)
* Added README to explain how to deal with (Arm CCA) reference values

paulhowardarm reviewed (approved):
Excellent enhancement - many thanks!
